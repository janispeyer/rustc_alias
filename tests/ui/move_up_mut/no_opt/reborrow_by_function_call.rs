// compile-flags: -Zmir-emit-retag

fn access(x: &mut i32) {
    *x = 5;
}

fn move_up_mut_no_opt(x: &mut i32, y: &mut i32) -> i32 {
    *x = 42;
    *y = 7;
    access(x);
    return *x;
}

fn main() {
    let mut x = 1;
    let mut y = 2;
    move_up_mut_no_opt(&mut x, &mut y);
}
