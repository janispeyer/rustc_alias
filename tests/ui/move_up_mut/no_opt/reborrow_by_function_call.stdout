# Analysing main
# Analysing reborrow_by_function_call
# MaybeTopOfBorrowStack Analysis
bb0
[ 0] [_1, _2] -> Retag([fn entry] _1)
[ 1] [_1, _2] -> Retag([fn entry] _2)
[ 2] [_1, _2] -> (*_1) = const 42_i32
[ 3] [_1, _2] -> (*_2) = const 7_i32
[ 4] [_1, _2] -> StorageLive(_4)
[ 5] [_1, _2] -> StorageLive(_5)
[ 6] [_2] -> _5 = &mut (*_1)
[ 7] [_2] -> Retag([2phase] _5)
[ 8] [_2] -> _4 = access(move _5) -> bb1

bb1
[ 0] [_2] -> StorageDead(_5)
[ 1] [_2] -> StorageDead(_4)
[ 2] [_2] -> _0 = (*_1)
[ 3] [_2] -> return

# FindImmutabilitySpans Analysis
bb0
[ 0] {} -> Retag([fn entry] _1)
[ 1] {} -> Retag([fn entry] _2)
[ 2] {_1: Span(bb0[2])} -> (*_1) = const 42_i32
[ 3] {_1: Span(bb0[2]), _2: Span(bb0[3])} -> (*_2) = const 7_i32
[ 4] {_1: Span(bb0[2]), _2: Span(bb0[3])} -> StorageLive(_4)
[ 5] {_1: Span(bb0[2]), _2: Span(bb0[3])} -> StorageLive(_5)
[ 6] {_1: Top, _2: Span(bb0[3])} -> _5 = &mut (*_1)
[ 7] {_1: Top, _2: Span(bb0[3])} -> Retag([2phase] _5)
[ 8] {_1: Top, _2: Span(bb0[3])} -> _4 = access(move _5) -> bb1

bb1
[ 0] {_1: Top, _2: Span(bb0[3])} -> StorageDead(_5)
[ 1] {_1: Top, _2: Span(bb0[3])} -> StorageDead(_4)
[ 2] {_1: Top, _2: Span(bb0[3])} -> _0 = (*_1)
[ 3] {_1: Top, _2: Span(bb0[3])} -> return


# CFG Before
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
(*_1) = const 42_i32
(*_2) = const 7_i32
StorageLive(_4)
StorageLive(_5)
_5 = &mut (*_1)
Retag([2phase] _5)
_4 = access(move _5) -> bb1

bb1
StorageDead(_5)
StorageDead(_4)
_0 = (*_1)
return

# CFG After
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
_6 = const 42_i32
(*_1) = _6
_7 = const 7_i32
(*_2) = _7
StorageLive(_4)
StorageLive(_5)
_5 = &mut (*_1)
Retag([2phase] _5)
_4 = access(move _5) -> bb1

bb1
StorageDead(_5)
StorageDead(_4)
_0 = (*_1)
return

# Analysing access
# MaybeTopOfBorrowStack Analysis
bb0
[ 0] [_1] -> Retag([fn entry] _1)
[ 1] [_1] -> (*_1) = const 5_i32
[ 2] [_1] -> _0 = const ()
[ 3] [_1] -> return

# FindImmutabilitySpans Analysis
bb0
[ 0] {} -> Retag([fn entry] _1)
[ 1] {_1: Span(bb0[1])} -> (*_1) = const 5_i32
[ 2] {_1: Span(bb0[1])} -> _0 = const ()
[ 3] {_1: Span(bb0[1])} -> return


# CFG Before
bb0
Retag([fn entry] _1)
(*_1) = const 5_i32
_0 = const ()
return

# CFG After
bb0
Retag([fn entry] _1)
_2 = const 5_i32
(*_1) = _2
_0 = const ()
return

